### PR TITLE
FIR: check subclass of sealed class

### PR DESCRIPTION
**Pull request by tgeng**
This PR reports the following

* SEALED_INHERITOR_IN_DIFFERENT_PACKAGE
* SEALED_INHERITOR_IN_DIFFERENT_MODULE
* CLASS_INHERITS_JAVA_SEALED_CLASS

Note that the current implementation reporting
SEALED_INHERITOR_IN_DIFFERENT_MODULE does not handle MPP yet.

### CODE REVIEW

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSubclassChecker.kt [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import org.jetbrains.kotlin.fir.analysis.diagnostics.DiagnosticReporter
+import org.jetbrains.kotlin.fir.analysis.diagnostics.FirErrors
+import org.jetbrains.kotlin.fir.analysis.diagnostics.reportOn
+import org.jetbrains.kotlin.fir.declarations.FirDeclarationOrigin
+import org.jetbrains.kotlin.fir.declarations.FirRegularClass
+import org.jetbrains.kotlin.fir.declarations.classId
+import org.jetbrains.kotlin.fir.declarations.isSealed
+import org.jetbrains.kotlin.fir.types.coneType
+
+object FirSealedSubclassChecker : FirRegularClassChecker() {
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSubclassChecker.kt:19]: What about `object`'s super types?

**Review by jsjeon — COMMENTED** [review]
Did you miss adding error messages?

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSubclassChecker.kt (hunk 1 above)
Comment (jsjeon) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSubclassChecker.kt:19]: Also, what about being part of `FirSealedSupertypeChecker`?

**Comment by tgeng** [comment]
Oops! Yep I missed the error message indeed!

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSubclassChecker.kt (hunk 1 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSubclassChecker.kt:19]: Anonymous object cannot subclass sealed class and hence these three checks are not applicable to that case.

Merging with `FirSealedSupertypeChecker` sounds good!

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt [hunk 2]
```diff
@@ -53,9 +73,10 @@ object FirSealedSupertypeChecker : FirClassChecker() {
                 continue
             }
 
-            val classSymbol = context.session.symbolProvider.getClassLikeSymbolByFqName(classId) as? FirRegularClassSymbol ?: continue
+            val fir = context.session.symbolProvider.getClassLikeSymbolByFqName(classId)
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt:76]: Nit: this change made the line shorter, so line wrapping seems not necessary.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt [hunk 3]
```diff
@@ -29,19 +31,37 @@ object FirSealedSupertypeChecker : FirClassChecker() {
...
+        for (superTypeRef in declaration.superTypeRefs) {
+            val classId = superTypeRef.coneType.classId ?: continue
 
             if (classId.isLocal) {
                 continue
             }
 
-            val classSymbol = context.session.symbolProvider.getClassLikeSymbolByFqName(classId) as? FirRegularClassSymbol ?: continue
+            val superClass = context.session.symbolProvider.getClassLikeSymbolByFqName(classId)
+                    as? FirRegularClassSymbol ?: continue
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt:44]: (Super) Nit: yeah, this is cosmetic (and surprised IJ formatter doesn't bother with this), but it looks to me quite unusual line wrapping.

https://developer.android.com/kotlin/style-guide#where_to_break

It might be better to glue type casting together, and rather break at `?:` I think.

**Review by jsjeon — APPROVED** [review]
Very minor comments. :)

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt (hunk 3 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt:44]: Thanks! Sounds good.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt [hunk 4]
```diff
@@ -29,19 +31,36 @@ object FirSealedSupertypeChecker : FirClassChecker() {
...
+            if (superClassPackage != subclassPackage) {
+                reporter.reportOn(
+                    superTypeRef.source,
+                    FirErrors.SEALED_INHERITOR_IN_DIFFERENT_PACKAGE,
+                    subclassPackage,
+                    superClassPackage,
+                    context
+                )
+            }
+            if (superClass.moduleData.session != subClassSession) {
```
Comment (demiurg906) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt:60]: Please compare module data instead of sessions. Module data was created specially for representing module which contains declaration

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt [hunk 5]
```diff
@@ -53,9 +72,9 @@ object FirSealedSupertypeChecker : FirClassChecker() {
                 continue
             }
 
-            val classSymbol = context.session.symbolProvider.getClassLikeSymbolByFqName(classId) as? FirRegularClassSymbol ?: continue
+            val fir = context.session.symbolProvider.getClassLikeSymbolByFqName(classId) as? FirRegularClassSymbol ?: continue
```
Comment (demiurg906) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt:75]: Property `fir` has `FirRegularClassSymbol` type. Can you please rename it back?

File: .idea/kotlinc.xml [hunk 6]
```diff
@@ -1,12 +1,5 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <project version="4">
-  <component name="Kotlin2JsCompilerArguments">
```
Comment (demiurg906) [inline .idea/kotlinc.xml:3]: Revert this change

**Review by demiurg906 — CHANGES_REQUESTED** [review]

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt [hunk 7]
```diff
@@ -29,19 +31,36 @@ object FirSealedSupertypeChecker : FirClassChecker() {
...
+            if (superClassPackage != subclassPackage) {
+                reporter.reportOn(
+                    superTypeRef.source,
+                    FirErrors.SEALED_INHERITOR_IN_DIFFERENT_PACKAGE,
+                    subclassPackage,
+                    superClassPackage,
+                    context
+                )
+            }
+            if (superClass.moduleData != subClassSession) {
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt:60]: Better to use `declaration.moduleData` directly here since it's only used here. Or at least rename `subClassSession` to `subclassModuleData` since it's not the session anymore 🙂

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt [hunk 8]
```diff
@@ -570,6 +573,15 @@ class FirDefaultErrorMessages {
             map.put(CLASS_IN_SUPERTYPE_FOR_ENUM, "Enum class cannot inherit from classes")
             map.put(SEALED_SUPERTYPE, "This type is sealed, so it can be inherited by only its own nested classes or objects")
             map.put(SEALED_SUPERTYPE_IN_LOCAL_CLASS, "Local class cannot extend a sealed class")
+            map.put(
+                SEALED_INHERITOR_IN_DIFFERENT_PACKAGE,
+                "Inheritor of sealed class or interface declared in package {0} but it must be in package {1} where base class is declared",
```
Comment (punzki) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt:578]: I was trying to see if the package names were in the right order in the message. But I found that the structure/grammar of this message is confusing and ambiguous. `{0}` could easily refer to the package of the sealed class or interface, and not the inheritor. Similarly, `{1}` could be interpreted as the package where the base class _is_ declared, not where it _should_ be.

I don't know if we are ok to change these messages, but I would even simplify it to `"A type can only inherit from a sealed class or interface if they are in the same package"`. Or if we absolutely must keep the package names: `"This supertype is a sealed class or interface in package {0} but the inheritor (currently in package {1}) must be in the same package"` and switch the package names around.

(Also "java" in the message below should be capitalized! But I see in FE 1.0 it is not.)

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt (hunk 7 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt:60]: 🤦

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt (hunk 8 above)
Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt:578]: OK, let me rephrase the message and put it into a separate commit so it they can opt to merge this separate commit or not. BTW, an interface cannot be sealed so I will drop that part.

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt [hunk 9]
```diff
@@ -29,19 +31,29 @@ object FirSealedSupertypeChecker : FirClassChecker() {
     }
 
     private fun checkGlobalDeclaration(declaration: FirClass, context: CheckerContext, reporter: DiagnosticReporter) {
-        for (it in declaration.superTypeRefs) {
-            val classId = it.coneType.classId ?: continue
+        val subclassPackage = declaration.classId.packageFqName
+        for (superTypeRef in declaration.superTypeRefs) {
+            val classId = superTypeRef.coneType.classId ?: continue
```
Comment (jsjeon) [inline compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/checkers/declaration/FirSealedSupertypeChecker.kt:36]: Super minor: if `classSymbol` below is renamed to `superClass`, isn't it aligned if we renamed this to `superClassId`? :)

File: compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt (hunk 8 above)
Comment (punzki) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt]: Looks good, but you should keep the interface part. Java has sealed interfaces (see `compiler/testData/diagnostics/tests/testsWithJava17/sealedClasses/kotlinInheritsJavaInterface.kt`). I just learned about this today too. 🤯

Comment (tgeng) [reply compiler/fir/checkers/src/org/jetbrains/kotlin/fir/analysis/diagnostics/FirDefaultErrorMessages.kt]: I don't think sealed Java interface is relevant here because it's an error to implement a sealed Java interface in Kotlin, which is why we have the `CLASS_INHERITS_JAVA_SEALED_CLASS` diagnostic devoted for this case.

**Comment by punzki** [comment]
Actually even in Kotlin, interfaces can be sealed:
https://kotlinlang.org/docs/sealed-classes.html

**Comment by tgeng** [comment]
> Actually even in Kotlin, interfaces can be sealed: https://kotlinlang.org/docs/sealed-classes.html

Thanks! I didn't know that! Updated.

**Comment by demiurg906** [comment]
Thank you. Merged manually
